src/hardware/voodoo.cpp

```diff
@@ -62,22 +63,21 @@ class VOODOO : public Module_base {
         UpdateConfiguration(configuration);
     }
 	
-    EmulationType GetEmulationTypeFromString(const std::string& voodoo_type_str) {
-        if (voodoo_type_str == "false") {
-            return EmulationType::None;
-        } else if ((voodoo_type_str == "software") || !OpenGL_using()) {
-            return EmulationType::Software;
-    #if C_OPENGL
-        } else if ((voodoo_type_str == "opengl") || (voodoo_type_str == "auto")) {
-            return EmulationType::OpenGL;
-    #else
-        } else if (voodoo_type_str == "auto") {
-            return EmulationType::Software;
-    #endif
-        } else {
-            return EmulationType::None;
-        }
-    }
+	EmulationType GetEmulationTypeFromString(const std::string& voodoo_type_str) {
+		static const std::unordered_map<std::string, EmulationType> str_to_emulation_type {
+			{"false", EmulationType::None},
+			{"software", EmulationType::Software},
+			{"opengl", C_OPENGL ? EmulationType::OpenGL : EmulationType::None},
+			{"auto", C_OPENGL ? EmulationType::OpenGL : EmulationType::Software}
+		};
+
+		auto it = str_to_emulation_type.find(voodoo_type_str);
+		if (it != str_to_emulation_type.end()) {
+			return it->second;
+		}
+
+		return !OpenGL_using() ? EmulationType::Software : EmulationType::None;
+	}
```

Comment:
This is far too complex just to convert strings to enum values.

If the things being looked-up were dynamic (like a database, or values were changing at runtime) - then a `std::map` is more suitable.

This is much too over-engineered for a fixed set of inputs and fixed set of outputs (and only four, at that).

Reply:
Would something like this be better?

```
constexpr EmulationType GetEmulationTypeFromString(const char* voodoo_type_str) {
    if (strcmp(voodoo_type_str, NONE_STR) == 0) {
        return EmulationType::None;
    } else if (strcmp(voodoo_type_str, SOFTWARE_STR) == 0) {
        return EmulationType::Software;
    } else if (strcmp(voodoo_type_str, OPENGL_STR) == 0) {
        return C_OPENGL ? EmulationType::OpenGL : EmulationType::None;
    } else if (strcmp(voodoo_type_str, AUTO_STR) == 0) {
        return C_OPENGL ? EmulationType::OpenGL : EmulationType::Software;
    } else {
        return !OpenGL_using() ? EmulationType::Software : EmulationType::None;
    }
}
```